pkg/apis/nmstate/v1alpha1/condition_types.go

```diff
@@ -0,0 +1,70 @@
+package v1alpha1
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ConditionList []Condition
+
+// +k8s:openapi-gen=true
+type Condition struct {
+	Type               ConditionType          `json:"type"`
+	Status             corev1.ConditionStatus `json:"status"`
+	Reason             ConditionReason        `json:"reason,omitempty"`
+	Message            string                 `json:"message,omitempty"`
+	LastHeartbeatTime  metav1.Time            `json:"lastHearbeatTime,omitempty"`
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+type ConditionType string
+
+// +k8s:openapi-gen=true
+type ConditionReason string
+
+func NewCondition(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) Condition {
+	now := metav1.Time{Time: time.Now()}
+	condition := Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastHeartbeatTime:  now,
+		LastTransitionTime: now,
+	}
+	return condition
+}
+
+func (conditions *ConditionList) Set(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) {
+	condition := conditions.Find(conditionType)
+
+	// If there isn't condition we want to change, add new one
+	if condition == nil {
+		condition := NewCondition(conditionType, status, reason, message)
+		*conditions = append(*conditions, condition)
+		return
+	}
+
+	now := metav1.Time{Time: time.Now()}
+
+	// If there is different status, reason or message update it
+	if condition.Status != status || condition.Reason != reason || condition.Message != message {
+		condition.Status = status
+		condition.Reason = reason
+		condition.Message = message
+		condition.LastTransitionTime = now
+	}
+	condition.LastHeartbeatTime = now
```

Comment:
Do we need to update the heartbeat even in case of no-op (i.e. we are passing the same exact condition in)?

Reply:
It's good to have it updated even for noop so you are sure that the status is being checked, and there is no issue with it.